File: tools/instant_casts.py
"""Instant casts: skills without an animation, found from what they leave in the log. Standard library only.

Applies Elite Insights' rules (tools/cast_rules.json, made by tools/gen_cast_rules.py) the way Elite
Insights does: each rule watches one kind of event (a buff gained, lost or given, an effect, a hit, a heal,
a missile, a minion spawning or casting), and every match by the same caster counts as one cast unless it
follows the previous match within the rule's ICD (50 ms by default).

Not applied: rules with custom C# check code, EngineerKitFinder, BandTogetherCastFinder and breakbar damage.

Usage:
    python tools/instant_casts.py <file.zevtc> [account]
"""
import collections
import json
import os
import struct
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from evtc import (Log, SC, TIME, SRC, DST, VALUE, BUFF_DMG, SKILL, SRC_INST, SRC_MASTER, DST_MASTER, IFF, BUFF,
                  RESULT, ACTIVATION, SHIELDS, OFFCYCLE, STATECHANGE, PROFESSIONS, spec_name)

HERE = os.path.dirname(os.path.abspath(__file__))
SERVER_DELAY = 10  # Elite Insights' ServerDelayConstant, ms
HEAL_SELF_REPORTED = 128

with open(os.path.join(HERE, "cast_rules.json"), encoding="utf-8") as f:
    _TABLE = json.load(f)
RULES = _TABLE["rules"]
NAMES = {int(k): v for k, v in _TABLE["names"].items()}  # Elite Insights' display names: use first
FALLBACK_NAMES = {int(k): v for k, v in _TABLE["fallback_names"].items()}  # use after the log's names
MINION_COMMAND_BUFF = _TABLE["minion_command_buff"]


class _Context:
    """The parts of a log the rules look at, indexed once."""

    def __init__(self, log):
        self.log = log
        ev = log.events
        self.build = next((e[SRC] for e in ev if e[STATECHANGE] == SC["GWBUILD"]), 0)
        guids = {e[SKILL]: struct.pack("<QQ", e[SRC], e[DST]).hex().upper()
                 for e in ev if e[STATECHANGE] == SC["IDTOGUID"]}

        # Minions: master from the master instance id their events carry
        inst_to_player = {}
        for e in ev:
            if e[STATECHANGE] == 0 and log.agents.get(e[SRC], {}).get("player"):
                inst_to_player.setdefault(e[SRC_INST], e[SRC])
        self.master = {}
        for e in ev:
            if e[SRC_MASTER] and e[SRC] not in self.master and not log.agents.get(e[SRC], {}).get("player"):
                if e[SRC_MASTER] in inst_to_player:
                    self.master[e[SRC]] = inst_to_player[e[SRC_MASTER]]
            if e[DST_MASTER] and e[DST] not in self.master and not log.agents.get(e[DST], {}).get("player"):
                if e[DST_MASTER] in inst_to_player:
                    self.master[e[DST]] = inst_to_player[e[DST_MASTER]]

        self.effects = collections.defaultdict(list)  # guid -> [(time, src, dst or None, duration)]
        agent_fx, ground_fx = SC["EFFECTAGENTCREATE"], SC["EFFECTGROUNDCREATE"]
        self.buff_apply = collections.defaultdict(list)  # buff -> [(time, by, to, duration)]
        self.buff_loss = collections.defaultdict(list)  # buff -> [(time, to)]
        self.damage = collections.defaultdict(list)  # skill -> [(time, from)]
        self.heals = collections.defaultdict(list)  # skill -> [(time, from, barrier)]
        self.missiles = collections.defaultdict(list)  # skill -> [(time, src)]
        self.minion_casts = collections.defaultdict(list)  # skill -> [(time, minion)]
        self.spawns = collections.defaultdict(list)  # agent -> [time]
        first_seen = {}
        for e in ev:
            sc, t = e[STATECHANGE], e[TIME]
            if sc == 0:
                self.damage[e[SKILL]].append((t, e[SRC]))
                first_seen.setdefault(e[SRC], t)
            elif sc in (agent_fx, ground_fx):
                guid = guids.get(e[SKILL])
                if guid:
                    duration = e[IFF] | e[BUFF] << 8 | e[RESULT] << 16 | e[ACTIVATION] << 24
                    self.effects[guid].append((t, e[SRC], e[DST] if sc == agent_fx else None, duration))
            elif sc == SC["BUFFAPPLY"]:
                self.buff_apply[e[SKILL]].append((t, e[SRC], e[DST], e[VALUE]))
            elif sc == SC["BUFFREMOVE_ALL"]:
                self.buff_loss[e[SKILL]].append((t, e[SRC]))
            elif sc == SC["EXTENSIONCOMBAT"] and e[OFFCYCLE] & HEAL_SELF_REPORTED:
                self.heals[e[SKILL]].append((t, e[SRC], bool(e[SHIELDS])))
            elif sc == SC["MISSILECREATE"]:
                self.missiles[e[SKILL]].append((t, e[SRC]))
            elif sc == SC["ANIMATIONSTART"] and e[SRC] in self.master:
                self.minion_casts[e[SKILL]].append((t, e[SRC]))
            elif sc == SC["SPAWN"]:
                self.spawns[e[SRC]].append(t)
        for a in self.master:
            if a not in self.spawns and a in first_seen:
                self.spawns[a].append(first_seen[a])
        self.has_effects = bool(self.effects)
        self.has_missiles = bool(self.missiles)

    def final_master(self, a):
        return self.master.get(a, a)

    def spec(self, a, base):
        ag = self.log.agents.get(a)
        if not ag or not ag["player"]:
            return None
        return PROFESSIONS.get(ag["prof"]) if base else spec_name(ag)

    def species(self, a):
        ag = self.log.agents.get(a)
        if not ag or ag["player"] or ag["prof"] >> 16 == 0xFFFF:
            return None
        return ag["prof"] & 0xFFFF


def _matches(rule, ctx):
    """Yield (time, caster) for every event the rule accepts, before the ICD."""
    kind, arg, minions = rule["type"], rule["arg"], rule["minions"]
    checks = rule["checks"]

    def spec_ok(who_agents):
        for c in checks:
            if c["kind"] != "spec":
                continue
            agent = who_agents.get(c["who"])
            spec = ctx.spec(agent, c["base"]) if agent is not None else None
            if (spec in c["specs"]) == c["negate"]:
                return False
        return True

    if kind in ("BuffGainCastFinder", "MinionCommandCastFinder", "BuffGiveCastFinder"):
        buff = MINION_COMMAND_BUFF if kind == "MinionCommandCastFinder" else arg
        for t, by, to, duration in ctx.buff_apply.get(buff, ()):
            if kind == "MinionCommandCastFinder" and (ctx.species(to) != arg or to not in ctx.master):
                continue
            durations = [c for c in checks if c["kind"] == "duration"]
            if durations and not any(abs(duration - c["values"][0]) < SERVER_DELAY for c in durations):
                continue
            if not spec_ok({"by": by, "to": to, "src": by, "dst": to}):
                continue
            key = by if kind == "BuffGiveCastFinder" else to
            yield t, ctx.final_master(key) if minions or kind != "BuffGainCastFinder" else key
    elif kind == "BuffLossCastFinder":
        for t, to in ctx.buff_loss.get(arg, ()):
            if spec_ok({"to": to, "dst": to}):
                yield t, ctx.final_master(to) if minions else to
    elif kind == "DamageCastFinder":
        for t, src in ctx.damage.get(arg, ()):
            yield t, src
    elif kind in ("EXTHealingCastFinder", "EXTBarrierCastFinder"):
        barrier = kind == "EXTBarrierCastFinder"
        for t, src, is_barrier in ctx.heals.get(arg, ()):
            if is_barrier == barrier:
                yield t, src
    elif kind in ("EffectCastFinder", "EffectCastFinderByDst"):
        by_dst = kind == "EffectCastFinderByDst"
        events = ctx.effects.get(arg, ())
        for t, src, dst, duration in events:
            key = dst if by_dst else src
            if key is None:
                continue
            ok = spec_ok({"src": src, "dst": dst})
            for c in checks:
                if not ok:
                    break
                if c["kind"] == "around_dst":
                    ok = (dst is not None) == c["want"]
                elif c["kind"] == "duration":
                    v = c["values"]
                    ok = duration == v[0] if len(v) == 1 else v[0] <= duration <= v[1]
                elif c["kind"] == "secondary_effect":
                    other = ctx.effects.get(c["guid"], ())
                    # a different effect (other GUID) by the same agent at the same moment; Elite Insights only
                    # excludes the event itself, which can't be in another GUID's list
                    ok = any((d if by_dst else s) == key and abs(t2 - c["offset"] - t) < SERVER_DELAY
                             for t2, s, d, _ in other)
            if ok:
                yield t, ctx.final_master(key) if minions else key
    elif kind == "MissileCastFinder":
        for t, src in ctx.missiles.get(arg, ()):
            yield t, ctx.final_master(src) if minions else src
    elif kind == "MinionCastCastFinder":
        for t, minion in ctx.minion_casts.get(arg, ()):
            yield t, ctx.final_master(minion)
    elif kind == "MinionSpawnCastFinder":
        wanted = set(arg) if isinstance(arg, list) else {arg}
        for minion, master in ctx.master.items():
            if ctx.species(minion) in wanted:
                for t in ctx.spawns.get(minion, ()):
                    yield t, master


SKIPPED_TYPES = {"EngineerKitFinder", "BandTogetherCastFinder", "BreakbarDamageCastFinder"}


def instant_casts(log):
    """{caster addr: {skill id: [times]}} for every player in the log."""
    ctx = _Context(log)
    out = collections.defaultdict(lambda: collections.defaultdict(list))
    for rule in RULES:
        if rule["custom"] or "unresolved" in rule or rule["type"] in SKIPPED_TYPES:
            continue
        if not rule["builds"][0] <= ctx.build < rule["builds"][1]:
            continue
        if rule.get("disabled_with") == "effect" and ctx.has_effects:
            continue
        if rule.get("disabled_with") == "missile" and ctx.has_missiles:
            continue
        last = collections.defaultdict(lambda: -10 ** 12)
        for t, caster in sorted(_matches(rule, ctx), key=lambda m: m[0]):
            if caster is None or not log.agents.get(caster, {}).get("player"):
                continue
            if t - last[caster] < rule["icd"]:
                last[caster] = t
                continue
            last[caster] = t
            out[caster][rule["skill"]].append(t + rule["time_offset"])
    return out


def main(path, account=None):
    log = Log(path)
    found = instant_casts(log)
    for caster, skills in found.items():
        ag = log.agents[caster]
        if account and ag["account"] != account:
            continue
        if not ag["account"]:
            continue
        line = ", ".join(f"{NAMES.get(s) or log.skills.get(s) or FALLBACK_NAMES.get(s, s)} {len(ts)}" for s, ts in
                         sorted(skills.items(), key=lambda kv: -len(kv[1])))
        print(f"{ag['name'][:24]:24} {spec_name(ag):13} {line}")


if __name__ == "__main__":
    main(*sys.argv[1:])

File: src/InstantCasts.cpp
#include "InstantCasts.h"

#include <algorithm>
#include <cstdio>
#include <cstring>
#include <string>
#include <unordered_set>

namespace InstantCasts
{
	namespace
	{
		using namespace Evtc;

		enum RuleType
		{
			CR_BuffGain, CR_BuffGive, CR_BuffLoss, CR_MinionCommand, CR_Damage, CR_EXTHealing, CR_EXTBarrier,
			CR_Effect, CR_EffectByDst, CR_Missile, CR_MinionCast, CR_MinionSpawn
		};

		struct CastRule
		{
			RuleType    Type;
			int32_t     Skill;
			int64_t     Arg;          // buff, skill, species
			const char* Guid;         // effect GUID (hex), for effect rules
			int         Species[4];   // minion spawn rules
			int64_t     Icd;
			uint64_t    BuildMin, BuildMax;
			bool        Minions;
			int64_t     TimeOffset;
			int         DisabledWith; // 1: when the log has effect data, 2: missile data
			const char* Spec;         // "who[!][^]=Spec|Spec;..."  ! negates, ^ base profession
			int64_t     DurationMin, DurationMax;
			int         AroundDst;    // 1: must be around an agent, 2: must not
			const char* SecondGuid;   // another effect by the same agent at the same time
			int64_t     SecondOffset;
		};

#include "CastRules.inc"

		constexpr int64_t kServerDelay = 10;
		constexpr uint8_t kHealSelfReported = 128;

		struct Effect { int64_t Time; uint64_t Src; uint64_t Dst; int64_t Duration; };
		struct BuffApply { int64_t Time; uint64_t By; uint64_t To; int64_t Duration; };

		struct Context
		{
			const Log* L = nullptr;
			uint64_t Build = 0;
			std::unordered_map<uint64_t, uint64_t> Master; // minion -> player
			std::unordered_map<std::string, std::vector<Effect>> Effects;
			std::unordered_map<uint32_t, std::vector<BuffApply>> Applies;
			std::unordered_map<uint32_t, std::vector<std::pair<int64_t, uint64_t>>> Losses, Damage, Missiles, MinionCasts;
			std::unordered_map<uint32_t, std::vector<std::tuple<int64_t, uint64_t, bool>>> Heals;
			std::unordered_map<uint64_t, std::vector<int64_t>> Spawns;

			uint64_t FinalMaster(uint64_t a) const { auto it = Master.find(a); return it == Master.end() ? a : it->second; }
			bool IsPlayer(uint64_t a) const { const Agent* ag = L->Find(a); return ag && ag->Player; }
			int Species(uint64_t a) const
			{
				const Agent* ag = L->Find(a);
				if (!ag || ag->Player || (ag->Prof >> 16) == 0xFFFF) { return -1; }
				return static_cast<int>(ag->Prof & 0xFFFF);
			}
			std::string Spec(uint64_t a, bool aBase) const
			{
				const Agent* ag = L->Find(a);
				if (!ag || !ag->Player) { return ""; }
				return aBase ? ProfessionName(ag->Prof) : SpecName(*ag);
			}
		};

		std::string GuidHex(uint64_t aLow, uint64_t aHigh)
		{
			unsigned char bytes[16];
			std::memcpy(bytes, &aLow, 8);
			std::memcpy(bytes + 8, &aHigh, 8);
			char out[33];
			for (int i = 0; i < 16; i++) { std::snprintf(out + 2 * i, 3, "%02X", bytes[i]); }
			return std::string(out, 32);
		}

		Context Build(const Log& aLog)
		{
			Context c;
			c.L = &aLog;
			std::unordered_map<uint16_t, uint64_t> instToPlayer;
			std::unordered_map<uint32_t, std::string> guids;
			for (const Event& e : aLog.Events)
			{
				if (e.StateChange == SC_GwBuild && !c.Build) { c.Build = e.Src; }
				else if (e.StateChange == SC_IdToGuid) { guids[e.Skill] = GuidHex(e.Src, e.Dst); }
				else if (e.StateChange == SC_None && c.IsPlayer(e.Src)) { instToPlayer.emplace(e.SrcInst, e.Src); }
			}
			for (const Event& e : aLog.Events)
			{
				if (e.SrcMaster && !c.Master.count(e.Src) && !c.IsPlayer(e.Src))
				{
					if (auto it = instToPlayer.find(e.SrcMaster); it != instToPlayer.end()) { c.Master[e.Src] = it->second; }
				}
				if (e.DstMaster && !c.Master.count(e.Dst) && !c.IsPlayer(e.Dst))
				{
					if (auto it = instToPlayer.find(e.DstMaster); it != instToPlayer.end()) { c.Master[e.Dst] = it->second; }
				}
			}
			std::unordered_map<uint64_t, int64_t> firstSeen;
			for (const Event& e : aLog.Events)
			{
				const int64_t t = static_cast<int64_t>(e.Time);
				switch (e.StateChange)
				{
				case SC_None:
					c.Damage[e.Skill].push_back({t, e.Src});
					firstSeen.emplace(e.Src, t);
					break;
				case SC_EffectAgentCreate: case SC_EffectGroundCreate:
				{
					auto it = guids.find(e.Skill);
					if (it == guids.end()) { break; }
					int64_t duration = e.Iff | (e.Buff << 8) | (e.Result << 16) | (static_cast<int64_t>(e.Activation) << 24);
					c.Effects[it->second].push_back({t, e.Src, e.StateChange == SC_EffectAgentCreate ? e.Dst : 0, duration});
					break;
				}
				case SC_BuffApply: c.Applies[e.Skill].push_back({t, e.Src, e.Dst, e.Value}); break;
				case SC_BuffRemoveAll: c.Losses[e.Skill].push_back({t, e.Src}); break;
				case SC_ExtensionCombat:
					if (e.Offcycle & kHealSelfReported) { c.Heals[e.Skill].push_back({t, e.Src, e.Shields != 0}); }
					break;
				case SC_MissileCreate: c.Missiles[e.Skill].push_back({t, e.Src}); break;
				case SC_AnimationStart: if (c.Master.count(e.Src)) { c.MinionCasts[e.Skill].push_back({t, e.Src}); } break;
				case SC_Spawn: c.Spawns[e.Src].push_back(t); break;
				default: break;
				}
			}
			for (auto& [minion, master] : c.Master)
			{
				if (!c.Spawns.count(minion)) { if (auto it = firstSeen.find(minion); it != firstSeen.end()) { c.Spawns[minion].push_back(it->second); } }
			}
			return c;
		}

		// "who[!][^]=A|B;..." against the agents playing each part
		bool SpecOk(const CastRule& r, const Context& c, uint64_t aSrc, uint64_t aDst)
		{
			if (!r.Spec[0]) { return true; }
			std::string all = r.Spec;
			size_t pos = 0;
			while (pos < all.size())
			{
				size_t end = all.find(';', pos);
				std::string part = all.substr(pos, end == std::string::npos ? std::string::npos : end - pos);
				pos = end == std::string::npos ? all.size() : end + 1;
				size_t eq = part.find('=');
				std::string who = part.substr(0, eq), specs = "|" + part.substr(eq + 1) + "|";
				bool negate = who.find('!') != std::string::npos, base = who.find('^') != std::string::npos;
				who.erase(std::remove_if(who.begin(), who.end(), [](char ch) { return ch == '!' || ch == '^'; }), who.end());
				uint64_t agent = who == "src" || who == "by" ? aSrc : aDst;
				std::string spec = agent ? c.Spec(agent, base) : "";
				bool in = !spec.empty() && specs.find("|" + spec + "|") != std::string::npos;
				if (in == negate) { return false; }
			}
			return true;
		}

		template <typename Emit>
		void Matches(const CastRule& r, const Context& c, Emit&& emit)
		{
			switch (r.Type)
			{
			case CR_BuffGain: case CR_BuffGive: case CR_MinionCommand:
			{
				uint32_t buff = r.Type == CR_MinionCommand ? kMinionCommandBuff : static_cast<uint32_t>(r.Arg);
				auto it = c.Applies.find(buff);
				if (it == c.Applies.end()) { return; }
				for (const BuffApply& a : it->second)
				{
					if (r.Type == CR_MinionCommand && (c.Species(a.To) != r.Arg || !c.Master.count(a.To))) { continue; }
					if (r.DurationMin >= 0 && std::llabs(a.Duration - r.DurationMin) >= kServerDelay) { continue; }
					if (!SpecOk(r, c, a.By, a.To)) { continue; }
					uint64_t key = r.Type == CR_BuffGive ? a.By : a.To;
					emit(a.Time, (r.Minions || r.Type != CR_BuffGain) ? c.FinalMaster(key) : key);
				}
				return;
			}
			case CR_BuffLoss:
				if (auto it = c.Losses.find(static_cast<uint32_t>(r.Arg)); it != c.Losses.end())
				{
					for (auto& [t, to] : it->second) { if (SpecOk(r, c, 0, to)) { emit(t, r.Minions ? c.FinalMaster(to) : to); } }
				}
				return;
			case CR_Damage:
				if (auto it = c.Damage.find(static_cast<uint32_t>(r.Arg)); it != c.Damage.end())
				{
					for (auto& [t, src] : it->second) { emit(t, src); }
				}
				return;
			case CR_EXTHealing: case CR_EXTBarrier:
				if (auto it = c.Heals.find(static_cast<uint32_t>(r.Arg)); it != c.Heals.end())
				{
					for (auto& [t, src, barrier] : it->second) { if (barrier == (r.Type == CR_EXTBarrier)) { emit(t, src); } }
				}
				return;
			case CR_Effect: case CR_EffectByDst:
			{
				auto it = c.Effects.find(r.Guid);
				if (it == c.Effects.end()) { return; }
				bool byDst = r.Type == CR_EffectByDst;
				for (const Effect& fx : it->second)
				{
					uint64_t key = byDst ? fx.Dst : fx.Src;
					if (!key || !SpecOk(r, c, fx.Src, fx.Dst)) { continue; }
					if (r.AroundDst && ((fx.Dst != 0) != (r.AroundDst == 1))) { continue; }
					if (r.DurationMin >= 0 && (fx.Duration < r.DurationMin || fx.Duration > r.DurationMax)) { continue; }
					if (r.SecondGuid[0])
					{
						auto other = c.Effects.find(r.SecondGuid);
						bool found = false;
						if (other != c.Effects.end())
						{
							for (const Effect& o : other->second)
							{
								if (&o == &fx) { continue; }
								if ((byDst ? o.Dst : o.Src) == key && std::llabs(o.Time - r.SecondOffset - fx.Time) < kServerDelay) { found = true; break; }
							}
						}
						if (!found) { continue; }
					}
					emit(fx.Time, r.Minions ? c.FinalMaster(key) : key);
				}
				return;
			}
			case CR_Missile:
				if (auto it = c.Missiles.find(static_cast<uint32_t>(r.Arg)); it != c.Missiles.end())
				{
					for (auto& [t, src] : it->second) { emit(t, r.Minions ? c.FinalMaster(src) : src); }
				}
				return;
			case CR_MinionCast:
				if (auto it = c.MinionCasts.find(static_cast<uint32_t>(r.Arg)); it != c.MinionCasts.end())
				{
					for (auto& [t, minion] : it->second) { emit(t, c.FinalMaster(minion)); }
				}
				return;
			case CR_MinionSpawn:
				for (auto& [minion, master] : c.Master)
				{
					int sp = c.Species(minion);
					if (sp < 0 || std::find(std::begin(r.Species), std::end(r.Species), sp) == std::end(r.Species) || sp == 0) { continue; }
					if (auto it = c.Spawns.find(minion); it != c.Spawns.end()) { for (int64_t t : it->second) { emit(t, master); } }
				}
				return;
			}
		}
	}

	std::unordered_map<uint64_t, std::vector<std::pair<int64_t, int32_t>>> Find(const Log& aLog)
	{
		Context c = Build(aLog);
		bool hasEffects = !c.Effects.empty(), hasMissiles = !c.Missiles.empty();
		std::unordered_map<uint64_t, std::vector<std::pair<int64_t, int32_t>>> out;
		std::vector<std::pair<int64_t, uint64_t>> found;
		for (const CastRule& r : kCastRules)
		{
			if (c.Build < r.BuildMin || c.Build >= r.BuildMax) { continue; }
			if ((r.DisabledWith == 1 && hasEffects) || (r.DisabledWith == 2 && hasMissiles)) { continue; }
			found.clear();
			Matches(r, c, [&](int64_t t, uint64_t caster) { found.push_back({t, caster}); });
			std::stable_sort(found.begin(), found.end(), [](auto& a, auto& b) { return a.first < b.first; });
			std::unordered_map<uint64_t, int64_t> last;
			for (auto& [t, caster] : found)
			{
				if (!caster || !c.IsPlayer(caster)) { continue; }
				auto it = last.find(caster);
				bool within = it != last.end() && t - it->second < r.Icd;
				last[caster] = t;
				if (!within) { out[caster].push_back({t + r.TimeOffset, r.Skill}); }
			}
		}
		return out;
	}

	const char* Name(int32_t aSkill, bool aFallback)
	{
		if (!aFallback) { for (auto& [id, name] : kSkillNames) { if (id == aSkill) { return name; } } }
		else { for (auto& [id, name] : kFallbackSkillNames) { if (id == aSkill) { return name; } } }
		return nullptr;
	}
}
